### PR TITLE
Redmine issue 1859: preparations to using SpecularSimulation in FitSuite

### PR DESCRIPTION
**Pull request by DmitryYurov**
Started working on creating IDetector. Some methods will be moved from IDetector2D to IDetector in later commits.

### CODE REVIEW

File: Core/Fitting/FitObject.cpp [hunk 1]
```diff
@@ -53,7 +53,7 @@ const OutputData<double>& FitObject::chiSquaredMap() const
     return *m_chi2_data.get();
 }
 
-const GISASSimulation& FitObject::simulation() const
+const Simulation& FitObject::simulation() const
```
Comment (gpospelov) [inline Core/Fitting/FitObject.cpp]: Can't find who uses the method. May be remove accessor on the next occasion?

File: Core/Fitting/FitSuite.cpp [hunk 2]
```diff
@@ -47,10 +47,25 @@ void FitSuite::addSimulationAndRealData(const Simulation& simulation,
                               const std::vector<std::vector<double>>& real_data,
                               double weight)
 {
+    // TODO: provide a way to construct OutputData<double> right from numpy array
     std::unique_ptr<IHistogram> data(IHistogram::createFrom(real_data));
-    addSimulationAndRealData(simulation, *data.get(), weight);
+    addSimulationAndRealData(simulation, *data, weight);
 }
 
+void FitSuite::addSimulationAndRealData(const Simulation& simulation,
```
Comment (gpospelov) [inline Core/Fitting/FitSuite.cpp]: May be on the next occasion create namespace OutputDataUtils and start moving similar methods there. Merge it also with IntensityDataFunctions.

File: Tests/UnitTests/Core/Fresnel/SpecularSimulationTest.h [hunk 3]
```diff
@@ -80,20 +85,46 @@ TEST_F(SpecularSimulationTest, ConstructSimulation)
...
-    EXPECT_EQ(size_t(10), sim.getScalarKz(0).size());
-
-    ASSERT_THROW( sim.getScalarR(3), std::runtime_error);
-    ASSERT_THROW( sim.getScalarT(3), std::runtime_error);
-    ASSERT_THROW( sim.getScalarKz(3), std::runtime_error);
+    EXPECT_EQ(10u, sim.getScalarR(0).size());
+    EXPECT_EQ(10u, sim.getScalarT(0).size());
+    EXPECT_EQ(10u, sim.getScalarKz(0).size());
+
+    Histogram1D* reflectivity = sim.reflectivity();
```
Comment (gpospelov) [inline Tests/UnitTests/Core/Fresnel/SpecularSimulationTest.h]: May be on the next occasion switch to smart pointers.

File: Core/Simulation/SpecularSimulation.cpp [hunk 4]
```diff
@@ -92,6 +92,36 @@ const IAxis* SpecularSimulation::getAlphaAxis() const
...
+        output[i] = std::norm((m_RT_coefficients[i][i_layer].get()->*fn_ptr)());
+    return output_ptr;
+}
+
+OutputData<double>* SpecularSimulation::getDetectorIntensity(IDetector2D::EAxesUnits) const
+{
+    return getData(0, &ILayerRTCoefficients::getScalarR).release();
+}
+
+Histogram1D* SpecularSimulation::reflectivity() const
```
Comment (gpospelov) [inline Core/Simulation/SpecularSimulation.cpp]: Please add methods to swig's Wrap/swig/libBornAgainCore.i similar to 
`%newobject GISASSimulation::getIntensityData`

File: Core/Export/ExportToPython.cpp [hunk 5]
```diff
@@ -73,9 +73,9 @@ namespace {
     //! Returns a function that converts a coordinate to a Python code snippet with appropiate unit
     std::function<std::string(double)> printFunc(const IDetector2D* detector)
     {
-        if (detector->getDefaultAxesUnits() == IDetector2D::MM)
+        if (detector->getDefaultAxesUnits() == DetectorAxesUnits::MM)
```
Comment (gpospelov) [inline Core/Export/ExportToPython.cpp]: I would even use simpler name AxesUnits on the next occasion.

File: Core/Instrument/IDetector2D.h [hunk 6]
```diff
@@ -100,16 +104,16 @@ class BA_CORE_API_ IDetector2D :  public IDetector
...
-    OutputData<double>* createDetectorMap(const Beam& beam, EAxesUnits units) const;
+    OutputData<double>* createDetectorMap(const Beam& beam, DetectorAxesUnits units) const;
 
     //! Returns vector of valid axes units
-    virtual std::vector<EAxesUnits> getValidAxesUnits() const;
+    virtual std::vector<DetectorAxesUnits> getValidAxesUnits() const;
 
     //! Return default axes units
-    virtual EAxesUnits getDefaultAxesUnits() const { return DEFAULT; }
+    virtual DetectorAxesUnits getDefaultAxesUnits() const { return DetectorAxesUnits::DEFAULT; }
```
Comment (gpospelov) [inline Core/Instrument/IDetector2D.h]: Such refactoring can be a nice opportunity to get rid of 'get' prefix in method names.

File: Core/Instrument/IDetector2D.h [hunk 7]
```diff
@@ -32,15 +32,20 @@ class IShape2D;
 class RegionOfInterest;
 class SimulationElement;
 
-//! Detector axis' units.
+//! Wrapper for detector axes units, required for a better representation of
+//! detector axes units in python
 //! @ingroup simulation
 
-enum class BA_CORE_API_ DetectorAxesUnits {DEFAULT, NBINS, RADIANS, DEGREES, MM, QYQZ};
+// workaround for SWIG (instead of just writing enum class DetectorAxesUnits...)
```
Comment (gpospelov) [inline Core/Instrument/IDetector2D.h]: We have to thing about moving swig-related code into some dedicated header at some point.

File: Core/Export/ExportToPython.cpp (hunk 5 above)
Comment (DmitryYurov) [reply Core/Export/ExportToPython.cpp]: Ok, will do it in next pull-request

File: Core/Fitting/FitObject.cpp (hunk 1 above)
Comment (DmitryYurov) [reply Core/Fitting/FitObject.cpp]: Done

File: Core/Instrument/IDetector2D.h (hunk 6 above)
Comment (DmitryYurov) [reply Core/Instrument/IDetector2D.h]: Ok, will do that in next pull-request

File: Core/Simulation/SpecularSimulation.cpp (hunk 4 above)
Comment (DmitryYurov) [reply Core/Simulation/SpecularSimulation.cpp]: Done

File: Tests/UnitTests/Core/Fresnel/SpecularSimulationTest.h (hunk 3 above)
Comment (DmitryYurov) [reply Tests/UnitTests/Core/Fresnel/SpecularSimulationTest.h]: Again, in next pull-request

**Pull request merged**